File: 108_Convert Sorted Array to Binary Search Tree.cpp
// Convert Sorted Array to Binary Search Tree
// Convert a sorted array into a height-balanced BST.
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{int val; TreeNode* left; TreeNode* right; TreeNode(int x): val(x), left(nullptr), right(nullptr) {}};

TreeNode* build(vector<int>& a, int l, int r){
    if(l>r) return nullptr;
    int mid = l + (r-l)/2;
    TreeNode* root = new TreeNode(a[mid]);
    root->left = build(a,l,mid-1);
    root->right = build(a,mid+1,r);
    return root;
}

TreeNode* sortedArrayToBST(vector<int>& nums){
    return build(nums, 0, nums.size()-1);
}

void inorder(TreeNode* r){ if(!r) return; inorder(r->left); cout<<r->val<<" "; inorder(r->right); }

int main(){
    vector<int> a = {-10,-3,0,5,9};
    TreeNode* root = sortedArrayToBST(a);
    cout<<"Inorder of BST: "; inorder(root); cout<<"\n";
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(log n) recursion

File: 108_Convert Sorted Array to Binary Search Tree.py
"""
Convert Sorted Array to Binary Search Tree
Build a height-balanced BST from a sorted array.
"""
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def sorted_array_to_bst(nums: List[int]) -> Optional[TreeNode]:
    def build(l, r):
        if l > r: return None
        mid = (l + r)//2
        node = TreeNode(nums[mid])
        node.left = build(l, mid-1)
        node.right = build(mid+1, r)
        return node
    return build(0, len(nums)-1)

def inorder(root):
    return inorder(root.left) + [root.val] + inorder(root.right) if root else []

if __name__ == '__main__':
    nums = [-10,-3,0,5,9]
    root = sorted_array_to_bst(nums)
    print('Inorder:', inorder(root))

# Time Complexity: O(n)
# Space Complexity: O(log n)
